Issue #559 - Do not reinstall already installed plugins

krakan opened the pull request:
This enhances downloading plugins when using a docker image that already has some pre-installed plugins.

Related: #559

krakan commented:
So, we're relying on this enhancement for our 70+ Jenkins instances. Any chance of having it merged?

oleg-nenashev commented:
Well, I recommend using @jenkinsci/code-reviewers when the review gets stuck

jglick commented:
FWIW this whole notion of using the update center to add plugins to a Jenkins image is pretty lame: too slow, and hard to make builds repeatable. In `workflow-aggregator-plugin/demo/` you can see a much faster and (IMO) more maintainable system. Unfortunately you do need a wrapper script (i.e., no “trusted builds”) to get caching benefits, since `docker-build` does not support mounting a reusable volume.

krakan commented:
Yes, sure, I could add a test scenario but I completely fail to run the tests though. The bats files start with
```
load 'test_helper/bats-support/load'
load 'test_helper/bats-assert/load'
```
and those files don't exist. Are they supposed to contain the definitions of eg. `assert_failure` and `assert_success`? Because if I just touch the `load.bash` files, the tests fail on missing those functions.

carlossg commented:
you need to clone with submodules, I'll add it to the readme

    git submodule update --init --recursive

krakan commented:
Ah, of course, submodules - my old nemesis ... ;)
Anyway, I added a test and changed some of the old ones to reflect changes in the output. My first test failed with some non-obvious reason. So I merged the test into the one that was already there and just changed an assert in that one.

krakan commented:
Is this OK now?
I guess it's a bit bad form to change an existing test but I believe that that test is actually wrong; verifying that we reinstall requested plugins doesn't mean we should reinstall dependencies too. Also that test would have failed already if there had actually been any bundled plugins.

carlossg commented on tests/install-plugins.bats:
```diff
@@ -84,8 +84,8 @@ load test_helpers
 @test "plugins are installed with install-plugins.sh even when already exist" {
   run docker build -t $SUT_IMAGE-install-plugins-update --no-cache $BATS_TEST_DIRNAME/install-plugins/update
   assert_success
+  assert_line --partial 'Skipping already installed dependency'
```
I think this would be clearer with the dependency name

    assert_line --partial 'Skipping already installed dependency: javadoc'
